javalin/src/main/java/io/javalin/plugin/PluginApi.kt

```diff
@@ -52,19 +52,8 @@ fun interface JavalinPlugin {
 
 }
 
-fun interface PluginFactory<PLUGIN : JavalinPlugin, CFG : PluginConfiguration> {
-
-    /**
-     * Create a new instance of the plugin with the given configuration.
-     */
-    fun create(config: Consumer<CFG>): PLUGIN
-
+abstract class Plugin<CONFIG>(userConfig: Consumer<CONFIG>? = null, defaultConfig: CONFIG) : JavalinPlugin {
+    protected val pluginConfig = defaultConfig.also { userConfig?.accept(it) } // apply user config to default config if present
 }
 
-/**
- * A marker interface for plugin configurations.
- */
-interface PluginConfiguration
-
-fun <CFG : PluginConfiguration> Consumer<CFG>.createUserConfig(cfg: CFG): CFG =
-    cfg.also { accept(it) }
+abstract class NoConfigPlugin : JavalinPlugin
```

Comment:
I don't like the fact that `JavalinPlugin` interface is implemented by some `abstract Plugin` and `abstract NoConfigPlugin`.
1. If we want to have both interface & abstract class, it should be called like `AbstractJavalinPlugin`/`StandardJavalinPlugin`/etc.
2. The fact that there's `JavalinPlugin` interface already removes the need to follow our requirements defined by the abstract constructor (and that was one of the main reasons to start this change 🤔)
3. Instead of making a 3rd structure that represents a no-config plugin, you could just make the config parameters optional, so `Plugin<Void>` already tells us that there's no cfg.

Reply:
1. Why should it be called that?
2. In this PR, `registerPlugin` accepts either a `Plugin` or a `NoConfigPlugin`, the `JavalinPlugin` interface is just used internally. The `registerPlugin` function does not accept `JavalinPlugin`.
3. That's how I had it, but it becomes very hacky. You need to cast a null to be a non-null (or have a nullable config), and you get a class that has a `pluginConfig` property that sometimes doesn't actually exist.

Reply:
One downside here is that people could potentially create their own NoConfigPlugin with a custom configuration 😓

Reply:
Maybe the hack is better.

Reply:
Added an exception to make it less hacky.